pkg/scheduler/framework/plugins/dynamicresources/structuredparameters.go

```diff
@@ -0,0 +1,235 @@
+		}
+		if claim.Status.Allocation == nil {
+			continue
+		}
+		for _, handle := range claim.Status.Allocation.ResourceHandles {
+			structured := handle.StructuredData
+			if structured == nil {
+				continue
+			}
+			if model[structured.NodeName] == nil {
+				model[structured.NodeName] = make(map[string]resourceModels)
+			}
+			resource := model[structured.NodeName][handle.DriverName]
+			for _, result := range structured.Results {
+				// Call AddAllocation for each known model. Each call itself needs to check for nil.
+				namedresourcesmodel.AddAllocation(&resource.namedresources, result.NamedResources)
+			}
+		}
+	}
+
+	return model, nil
+}
+
+func newClaimController(logger klog.Logger, class *resourcev1alpha2.ResourceClass, classParameters *resourcev1alpha2.ResourceClassParameters, claimParameters *resourcev1alpha2.ResourceClaimParameters) (*claimController, error) {
+	// Each node driver is separate from the others. Each driver may have
+	// multiple requests which need to be allocated together, so here
+	// we have to collect them per model.
+	type perDriverRequests struct {
+		parameters []runtime.RawExtension
+		requests   []*resourcev1alpha2.NamedResourcesRequest
+	}
+	namedresourcesRequests := make(map[string]perDriverRequests)
+	for i, request := range claimParameters.DriverRequests {
+		driverName := request.DriverName
+		p := namedresourcesRequests[driverName]
+		for e, request := range request.Requests {
+			switch {
+			case request.ResourceRequestModel.NamedResources != nil:
+				p.parameters = append(p.parameters, request.VendorParameters)
+				p.requests = append(p.requests, request.ResourceRequestModel.NamedResources)
+			default:
+				return nil, fmt.Errorf("claim parameters %s: driverRequersts[%d].requests[%d]: no supported structured parameters found", klog.KObj(claimParameters), i, e)
+			}
+		}
+		if len(p.requests) > 0 {
+			namedresourcesRequests[driverName] = p
+		}
+	}
+
+	c := &claimController{
+		class:           class,
+		classParameters: classParameters,
+		claimParameters: claimParameters,
+		namedresources:  make(map[string]perDriverController, len(namedresourcesRequests)),
+	}
+	for driverName, perDriver := range namedresourcesRequests {
+		var filter *resourcev1alpha2.NamedResourcesFilter
+		if classParameters != nil {
+			for _, f := range classParameters.Filters {
+				if f.DriverName == driverName && f.ResourceFilterModel.NamedResources != nil {
+					filter = f.ResourceFilterModel.NamedResources
+					break
+				}
+			}
+		}
+		controller, err := namedresourcesmodel.NewClaimController(filter, perDriver.requests)
+		if err != nil {
+			return nil, fmt.Errorf("creating claim controller for named resources structured model: %w", err)
+		}
+		c.namedresources[driverName] = perDriverController{
+			parameters: perDriver.parameters,
+			controller: controller,
+		}
+	}
+	return c, nil
+}
+
+// claimController currently wraps exactly one structured parameter model.
```

Comment:
can we make a better name as "controller" in the context of k8s is usually perceived as a component doing endless reconcilation..

Reply:
It's named like this because this matches the control plane "DRA driver controller" which reconciles claim status. One could argue that this in-tree controller is also reconciling resource requests with resource claims, and it affects the claim status, so perhaps the name fits?

Reply:
Having said that, it's also different. I'm open for suggestions.